File: LinkedList/SkipMDeleteN.py
# Python program to skip M nodes and then delete N nodes alternately in a singly linked list


# Node class
class Node:

    # Constructor to initialise data and next
    def __init__(self, data=None):
        self.data = data
        self.next = None


class SinglyLinkedList:

    # Constructor to initialise head
    def __init__(self):
        self.head = None

    # Function to skip M delete N nodes
    def skip_m_delete_n(self, m, n):
        current = self.head

        # if list is empty return
        if current is None:
            return

        # Main loop to traverse the whole list
        while current is not None:
            # loop to skip M nodes
            for i in range(1, m):
                if current is None:
                    return
                current = current.next

            if current is None:
                return

            # loop to delete N nodes
            temp = current.next
            for i in range(1, n+1):
                if temp is None:
                    break
                temp = temp.next

            # Point the last node skipped to the node after N nodes deletion
            current.next = temp
            # set current for next iteration
            current = temp

    # Function to Insert data at the beginning of the linked list
    def insert_at_beg(self, data):
        node = Node(data)
        node.next = self.head
        self.head = node

    # Function to print the linked list
    def print_data(self):
        current = self.head
        while current is not None:
            print(current.data, '-> ', end='')
            current = current.next
        print('None')

if __name__ == '__main__':
    linked_list = SinglyLinkedList()
    linked_list.insert_at_beg(9)
    linked_list.insert_at_beg(8)
    linked_list.insert_at_beg(7)
    linked_list.insert_at_beg(6)
    linked_list.insert_at_beg(5)
    linked_list.insert_at_beg(4)
    linked_list.insert_at_beg(3)
    linked_list.insert_at_beg(2)
    linked_list.insert_at_beg(1)

    linked_list.print_data()

    # call the skip_m_delete_n function
    linked_list.skip_m_delete_n(2, 2)

    # print the modified linked list
    linked_list.print_data()

File: LinkedList/SkipMDeleteN.cpp
// Skip M nodes and delete N nodes from a linked list alternately

#include <iostream>

using namespace std;

class Node {
    public:
        int data;
        Node* next;

        Node() {
            this->next = NULL;
        }

        Node(int data) {
            this->data = data;
            this->next = NULL;
        }
};

class LinkedList {
    public:
        Node* head = NULL;

        void insert_at_beg(int data){
            Node* temp = new Node(data);
            temp->next = this->head;
            this->head = temp;
        }

        void print_data(){
            Node* itr = this->head;

            while(itr != NULL){
                cout << itr->data << " ";
                itr = itr->next;
            }
            cout << endl;
        }
};

void skip_m_delete_n(LinkedList* ll, int m, int n){
    Node *curr = ll->head, *del_ptr, *temp;

    while(curr != NULL){
        // Skip M nodes, reach to the mth node
        for(int i=0; i < m-1 && curr != NULL; i++){
            curr = curr->next;
        }
        if(curr == NULL) return;

        del_ptr = curr->next;

        // Delete N nodes, reach to the node and replace pointers
        for(int i=0; i < n && del_ptr != NULL; i++){
            temp = del_ptr;
            del_ptr = del_ptr->next;
            free(temp);
        }
        
        curr->next = del_ptr;
        curr = curr->next;
    }
}

int main() {
    LinkedList* ll = new LinkedList();
    
    // Create Linked list
    ll->insert_at_beg(8);
    ll->insert_at_beg(7);
    ll->insert_at_beg(6);
    ll->insert_at_beg(5);
    ll->insert_at_beg(4);
    ll->insert_at_beg(3);
    ll->insert_at_beg(2);
    ll->insert_at_beg(1);

    cout << "Linked list before removal: " << endl;
    ll->print_data();
    int k = 3;
    int m = 2, n = 3;
    skip_m_delete_n(ll, m, n);
    cout << "Linked list after skip "<< m << " delete " << n << ": "<< endl;
    ll->print_data();
    return 0;
}


